services/app-api/src/postgres/contractAndRates/replaceRateOnContract.ts

```diff
@@ -0,0 +1,209 @@
+    }
+    const updateRates: UpdateDraftContractRatesArgsType['rateUpdates']['update'] =
+        []
+    const linkRates: UpdateDraftContractRatesArgsType['rateUpdates']['link'] =
+        []
+
+    previousRates.forEach((rate, idx) => {
+        if (rate.id === withdrawnRateID) {
+            return // we already know this is swapped, we will swap in replacement later, skip for now
+        }
+        // keep any existing linked rates besides replacement or withddrawn rate
+        if (rate.parentContractID !== contractID) {
+            linkRates.push({
+                rateID: rate.id,
+                ratePosition: idx + 1,
+            })
+        } else {
+            // keep any existing child rates
+            updateRates.push({
+                rateID: rate.id,
+                formData:
+                    rate.packageSubmissions[0].contractRevisions[0].formData,
+                ratePosition: idx + 1,
+            })
+        }
+    })
+    // add in replacement rate with the ratePosition of the withdrawn rate
+    linkRates.push({
+        rateID: replacementRateID,
+        ratePosition: withdrawnRateIndex + 1,
+    })
+
+    // arrange data for update draft contract rates
+    const updateContractRatesArgs: UpdateDraftContractRatesArgsType = {
+        contractID,
+        rateUpdates: {
+            create: [],
+            update: [
+                //  keep any already added child rates
+                ...updateRates,
+            ],
+            delete: [],
+            unlink: [
+                // unlink the withdrawn child rate
+                {
+                    rateID: withdrawnRateID,
+                },
+            ],
+            link: [
+                // link replacement rate and keep any other already added linked rates
+                ...linkRates.sort((a, b) => a.ratePosition - b.ratePosition),
+            ],
+        },
+    }
+
+    const updateResult = await updateDraftContractRatesInsideTransaction(
+        tx,
+        updateContractRatesArgs
+    )
+    if (updateResult instanceof Error) {
+        return updateResult
+    }
+
+    // resubmit contract
+    const resubmitContractArgs: SubmitContractArgsType = {
+        contractID,
+        submittedByUserID: replacedByUserID,
+        submittedReason: replaceReason,
+    }
+    const resubmitResult = await submitContractInsideTransaction(
+        tx,
+        resubmitContractArgs
+    )
+    if (resubmitResult instanceof Error) {
+        return resubmitResult
+    }
+
+    // finalize withdraw rate on the child rate we just removed from the contract
+    const withdrawRateArgs: WithdrawDateArgsType = {
+        rateID: withdrawnRateID,
+        withdrawnByUserID: replacedByUserID,
+        withdrawReason: replaceReason,
+    }
+    const withdrawResult = await withdrawRateInsideTransaction(
+        tx,
+        withdrawRateArgs
+    )
+    if (withdrawResult instanceof Error) {
+        return withdrawResult
+    }
+    return updateResult
```

Comment:
Great job with the logic in this file! I follow everything except, what's the reasoning for why is `updateResult` returned instead of `resubmitResult`?

Reply:
@pearl-truss Added a comment about this `return updateResult because we want to return updated contract - withdrawResult is the updated standalone rate`